pkgs/build-support/rust/lib/default.nix

```diff
@@ -23,8 +23,9 @@ rec {
 
     ccForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
-    linkerForHost = if shouldUseLLD stdenv.targetPlatform
-      && !stdenv.cc.bintools.isLLVM
+    # For wasi: rustc passes lld-flavored args unless the compiler is named "clang". So use lld.
```

Comment:
Do you know which code in rustc does this check?

Reply:
After some digging. I think I found the [relevant part](https://github.com/rust-lang/rust/blob/adeb79d3f523dfc11a8e9db64ddfc5f546f650bb/compiler/rustc_target/src/spec/mod.rs#L314-L324). So `clang` or `gcc` as names would be recognized, but not `cc` (as set in `ccForHost`). I've revised the comment accordingly.
(Defaulting to lld without being wrapped in a compiler is [here](https://github.com/rust-lang/rust/blob/adeb79d3f523dfc11a8e9db64ddfc5f546f650bb/compiler/rustc_target/src/spec/base/wasm.rs#L90).)

Side notes:
* Since `cc` isn't recognized as anything by rust, the current setup for `ccForHost` is maybe not the best in general.
* [This doccoment](https://doc.rust-lang.org/beta/nightly-rustc/rustc_target/spec/enum.LinkerFlavor.html) has some helpful insights.

Reply:
> * Since `cc` isn't recognized as anything by rust, the current setup for `ccForHost` is maybe not the best in general…?

Yeah, that's what I was wondering.

Reply:
It might also make sense to teach Rust to understand "cc", as it's not exactly a Nixpkgs-specific convention.

Reply:
I think I read somewhere that not recognizing cc is intentional. Somebody felt that two letters is too little to go on? I might be mistaken, my memory on this is rather hazy. (In any case, this is a finicky part of rustc. Just looking at the amount of "fix" [here](https://github.com/rust-lang/rust/blame/498553fc04f6a3fdc53412320f4e913bc53bc267/compiler/rustc_codegen_ssa/src/back/link.rs#L1315-L1322). I'd rather not touch it.)

In any case, that's not something I intend to fix here. ^^

Reply:
There are other options?

In case you're thinking of just doing
```nix
    linkerForHost = if stdenv.cc.bintools.isLLVM
      then "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}clang"
      else "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}gcc";
    # resp. doing this for all of cc/cxx/linkerForHost
```

that won't work without further adjustments either:

```
  = note: clang: error: unable to execute command: Executable "wasm-ld" doesn't exist!
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

And I really don't now what those adjustments are, or whether they're benign. I've looked into whether clang can somehow be taught a different linker command before, but couldn't find out anything.

(This would also severely increase the blast radius of this PR…)

Reply:
I dug around a bit (for options), here's a few things I found:

1. It's unlikely that rustc upstream will accept patches that recognize a `-cc` suffix as a compiler wrapping a linker, reason being that `-cc` might be anything, so it's unclear what arguments can be passed to it
2. cargo-auditable passes linker flags with `-Wl,` unconditionally ([src](https://github.com/rust-secure-code/cargo-auditable/blob/38b37332dfa3601f90608c1565fbd6e63bd6b08a/cargo-auditable/src/rustc_wrapper.rs#L64)) meaning we can't just use a plain linker as `linkerForHost`, it has to be invoked through a compiler as long as we want to use cargo-auditable.
3. 1 and 2 combined mean that some `if stdenv.cc.bintools.isLLVM then clang else gcc` is necessary for `linkerFor…`.
4. Using clang as a linker may work, but it's mechanism for finding wasm-ld is not nice: It searches paths passed with `-B` for `<triple>-wasm-ld` and `wasm-ld` ([src](https://github.com/llvm/llvm-project/blob/ad7aeb0ff58ebd29f68adb85c64e8010639e2a76/clang/lib/Driver/Driver.cpp#L6219-L6226)). However, it doesn't properly normalize the triple, and just uses the `-target wasm32-wasi` argument it gets from rustc as is. Clang itself doesn't think that's a good state and does have a [TODO](https://github.com/llvm/llvm-project/blob/ad7aeb0ff58ebd29f68adb85c64e8010639e2a76/clang/lib/Driver/Driver.cpp#L6187) about this. 
  (This might also be fixable in rustc, which does seem to have logic to not directly pass its own target name. But I havent looked into how that works.)
5. The way the linker gets passed from bintools to clang is:
   * LLVM bintools gets compiled
   * bintools-wrapper wraps it (currently not including wasm-ld) ([src](https://github.com/NixOS/nixpkgs/blob/999be9b3390bbff39f6adb9335f215c2b035744a/pkgs/build-support/bintools-wrapper/default.nix#L242-L245))
   * cc-wrapper symlinks that into ints own bin directory ([src](https://github.com/NixOS/nixpkgs/blob/f8b69469ae63cd405b22bff2a949ba5721a21017/pkgs/build-support/cc-wrapper/default.nix#L371-L374))
   * cc-wrapper's `add-flags.sh` passes the `-B` flag to its own bin dir ([src](https://github.com/NixOS/nixpkgs/blob/d1ae288c68a3a3455e46cbd7c35c0babb3485c79/pkgs/build-support/cc-wrapper/add-flags.sh#L30-L31))
6. 4+5 could in theory be worked around like [this](https://github.com/user-attachments/files/16284747/changerustlinker.patch.txt), but that's a far more terrible hack than what I've tried to do here.

So, the only way I see to get this to work cleanly is to submit a fix to clang for triple normalization, and then submit a PR to nixpkgs staging that switches to `if stdenv.cc.bintools.isLLVM then clang else gcc` (oh the rebuilds, I don't have hardware to test this), and then finally get back to this. I think that's above the level of difficulty I can manage.  

If you've got any advice on how else I could approach this, please do tell.